### PR TITLE
feat(external): implement METHLYVI for scBS-seq data

### PR DESCRIPTION
**Pull request by ethanweinberger**
This PR implements methylation variational inference (methylVI) for modeling methylation data from single-cell bisulfite sequencing (scBS-seq) experiments, described originally in [Weinberger and Lee (2024).](https://openreview.net/forum?id=Mg2DM0F3AY). In short, methylVI assumes that counts of methylated cytosines across pre-defined genomic region features follow a [Beta-Binomial distribution](https://docs.scvi-tools.org/en/latest/api/reference/scvi.distributions.BetaBinomial.html). As a placeholder citation I put in a reference to the workshop paper with the intention of swapping in a citation to the full paper once we upload it to bioRxiv.

All implementation code can be found in `scvi/external/methylvi`. Of note, as part of this PR I've added a new `DecoderMETHYLVI` class in `external/methylvi/_base_components.py` for obtaining decoded Beta-Binomial parameters for BS-seq data. This class may be useful for the development of future models, and might be worth moving out of external. Otherwise, the bulk of the implementation code can be found in `scvi/external/methylvi/_model.py` and `scvi/external/methylvi/_module.py`.

I structured my implementation of methylVI so that the model could be used with a single methylation modality an AnnData object (e.g. CpG gene body features) or multiple modalities in a MuData object (e.g. both CpG and non-CpG gene body features as separate modalities). To avoid repeating too much code, under the hood I tried to treat models trained with AnnData's as if they were trained using a MuData with a single modality (see my comments for more details).

@martinkim0 @canergen for testing I added some basic sanity check tests (loosely based on the tests in `tests/model/test_peakvi.py`) confirming that (1) the model trains without errors and (2) downstream analysis functions (e.g. `get_latent_representation`) run without errors. Let me know if there are additional tests you'd like to see.

### CODE REVIEW

File: src/scvi/external/methylvi/_model.py [hunk 1]
```diff
@@ -0,0 +1,704 @@
...
+                    mod_key=mod,
+                    is_count_data=True,
+                    mod_required=True,
+                )
+            )
+
+        mudata_fields = mc_fields + cov_fields + [batch_field]
+        adata_manager = AnnDataManager(fields=mudata_fields, setup_method_args=setup_method_args)
+        adata_manager.register_fields(mdata, **kwargs)
+        adata_manager.modalities = methylation_modalities
```
Comment (ethanweinberger) [inline src/scvi/external/methylvi/_model.py:256]: This (and a similar line in `setup_anndata`) was a hack so that I could access modality names later when instantiating the class. Is there a smarter way to do this?

File: docs/user_guide/models/methylvi.md [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+-   Scalable to very large datasets (>1 million cells).
+
+The limitations of methylVI include:
+
+-   Effectively requires a GPU for fast inference.
+-   Latent space is not interpretable, unlike that of a linear method.
+
+```{topic} Tutorials:
+
+-   Work in progress.
```
Comment (ethanweinberger) [inline docs/user_guide/models/methylvi.md:18]: As soon as the model is merged into scvi-tools I can add tutorial notebooks from [our original repository](https://github.com/suinleelab/methylVI) to the scvi-tools repo.

**Review by martinkim0 — COMMENTED** [review]
Hey, sorry for the delay in reviewing this. Left some minor comments about code style. One major comment: how would you feel about removing the `setup_anndata` method in favor of just having `setup_mudata`? This way we don't have to rely on the current modality placeholder code.

File: src/scvi/model/_scvi.py [hunk 3]
```diff
@@ -91,7 +91,7 @@ class SCVI(
     >>> vae = scvi.model.SCVI(adata)
     >>> vae.train()
     >>> adata.obsm["X_scVI"] = vae.get_latent_representation()
-    >>> adata.obsm["X_normalized_scVI"] = vae.get_normalized_expression()
+    >>> adata.obsm["X_normalized_scVI"] = vae.get_normalized_methylation()
```
[the remark this answers is not in the document]
Comment (ethanweinberger) [reply src/scvi/model/_scvi.py]: Thanks for the catch- fixed!

File: src/scvi/external/methylvi/_model.py [hunk 4]
```diff
@@ -0,0 +1,704 @@
...
+from scvi.utils import setup_anndata_dsp
+
+from ._constants import METHYLVI_REGISTRY_KEYS
+from ._module import METHYLVAE
+from ._utils import scmc_raw_counts_properties
+
+logger = logging.getLogger(__name__)
+
+
+class MethylVIModel(VAEMixin, UnsupervisedTrainingMixin, ArchesMixin, BaseModelClass):
```
[the remark this answers is not in the document]
Comment (ethanweinberger) [reply src/scvi/external/methylvi/_model.py]: Makes sense to me. Done!

File: src/scvi/external/methylvi/_model.py [hunk 5]
```diff
@@ -0,0 +1,704 @@
...
+    """
+    Model class for methylVI
```
[the remark this answers is not in the document]
Comment (ethanweinberger) [reply src/scvi/external/methylvi/_model.py]: Done. Thanks for the tip!

**Comment by ethanweinberger** [comment]
Hi @martinkim0 @canergen. Apologies for the delay in responding to the feedback. 

@martinkim0 I accepted all of your style-related suggestions. Regarding your major comment on the placeholder modality code, I went ahead with your suggestion to have the model only support MuData objects. Please let me know if you have more feedback or if I missed anything on my end!

File: docs/user_guide/models/methylvi.md [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+This is achieved using the method:
+
+```
+>>> adata.obsm["X_methylvi"] = model.get_latent_representation()
+```
+
+Users may also return samples from this distribution, as opposed to the mean, by passing the argument `give_mean=False`.
+The latent representation can be used to create a nearest neighbor graph with scanpy with:
+
+```
```
[the remark this answers is not in the document]
Comment (ethanweinberger) [reply docs/user_guide/models/methylvi.md:129]: When I attempted to implement this change, I ran into issues with precommit complaining about not being able to parse the subsequent code block. Is there anything else I need to do here?

File: src/scvi/external/methylvi/_model.py [hunk 7]
```diff
@@ -0,0 +1,627 @@
...
+
+    @classmethod
+    @setup_anndata_dsp.dedent
+    def setup_mudata(
+        cls,
+        mdata: MuData,
+        mc_layer: str,
+        cov_layer: str,
+        methylation_modalities: Iterable[str],
+        batch_key: str | None = None,
```
Comment (canergen) [inline src/scvi/external/methylvi/_model.py:136]: Can you add support for categorical covariates? We will deprecate batch_key soon'ish as it's confusing for users (difference between batch_key and categorical_covariate_key).

File: src/scvi/external/methylvi/_model.py [hunk 8]
```diff
@@ -0,0 +1,627 @@
...
+        %(de_mode)s
+        %(de_delta)s
+        %(de_batch_size)s
+        %(de_all_stats)s
+        %(de_batch_correction)s
+        %(de_batchid1)s
+        %(de_batchid2)s
+        %(de_fdr_target)s
+        %(de_silent)s
+        two_sided
```
Comment (canergen) [inline src/scvi/external/methylvi/_model.py:547]: We are currently adding this to the full DE code (three ways - smaller, larger, equal). I assume it's different here as you only test for larger and not for smaller?

File: src/scvi/external/methylvi/_module.py [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+        return outputs
+
+    @auto_move_data
+    def generative(self, z, batch_index):
+        """Runs the generative model."""
+        # form the parameters of the BetaBinomial likelihood
+        px_mu, px_gamma = {}, {}
+        for modality in self.modalities:
+            px_mu[modality], px_gamma[modality] = self.decoders[modality](
+                self.dispersion, z, batch_index
```
Comment (canergen) [inline src/scvi/external/methylvi/_module.py:163]: Can you explain the modalities? And why they aren't fed to the encoder? Is it for different technologies? Are the cells with multiple modalities?

File: src/scvi/external/methylvi/_module.py [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+                }
+            )
+
+    def _get_inference_input(self, tensors):
+        """Parse the dictionary to get appropriate args"""
+        mc = torch.cat(
+            [tensors[f"{x}_{METHYLVI_REGISTRY_KEYS.MC_KEY}"] for x in sorted(self.modalities)],
+            dim=1,
+        )
+        cov = torch.cat(
```
Comment (canergen) [inline src/scvi/external/methylvi/_module.py:111]: I assumed they can have different nu,ber of features. Why does it work here?

Comment (canergen) [reply src/scvi/external/methylvi/_module.py:111]: If they all have the same shape, and there is only one modality per cell. Shouldn't we ask the user to concatenate before inputing to methylVI. I assume those things are clear but I'm struggling with the meaning of modality.

File: src/scvi/external/methylvi/_utils.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+        subset of variables to extract properties from. if None, all variables are used.
+
+    Returns
+    -------
+    type
+        Dict of ``np.ndarray`` containing, by pair (one for each sub-population).
+    """
+    adata = adata_manager.adata
+
+    if isinstance(adata, MuData):
```
Comment (canergen) [inline src/scvi/external/methylvi/_utils.py:43]: Do we still need this one, if we require MuData input?

File: src/scvi/external/methylvi/_model.py [hunk 12]
```diff
@@ -0,0 +1,627 @@
...
+                generative_kwargs={},
+                compute_loss=False,
+            )
+
+            for modality in self.modalities:
+                exp_ = generative_outputs["px_mu"][modality]
+                exp_ = exp_[..., region_mask]
+                exprs[modality].append(exp_.cpu())
+            if store_distributions:
+                qz_store.store_distribution(inference_outputs["qz"])
```
Comment (canergen) [inline src/scvi/external/methylvi/_model.py:371]: Are qz_store and px_store for all reasonable cases of modalities?

**Comment by canergen** [comment]
Hi Ethan, taking over from Martin here. Can you help me with the definition of modalities. See my comments above. The rest LGTM.

**Comment by canergen** [comment]
My suggestion is that we will merge it with the biorxiv version and the tutorials?

File: src/scvi/external/methylvi/_utils.py (hunk 11 above)
Comment (ethanweinberger) [reply src/scvi/external/methylvi/_utils.py]: Thanks for the catch. Fixed!

**Comment by ethanweinberger** [comment]
> Hi Ethan, taking over from Martin here. Can you help me with the definition of modalities. See my comments above. The rest LGTM.

Hey Can. Thanks for the feedback!

To clarify, a modality here represents a combination of cytosine context (e.g. CpG versus non-CpG) and sequence context (e.g. gene bodies versus promoter regions versus enhancer regions etc.). This interpretation of modality is similar to that previously used in the methylation section of the [MOFA+ paper](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02015-1) . While the same Beta-Binomial likelihood is used for any of these combinations, I think it makes more sense to organize the data in MuData format (and have methylVI figure out how to combine the inputs) rather than asking the user to concatenate everything manually. My reasons are:

1. I think would be cumbersome + easy to make mistakes for users if they need to concatenate features manually for model training and then potentially split them during downstream analyses. For example, users may want to train a model using multiple methylation modalities, but then analyze differential methylation results and e.g. perform pathway analysis separately for CpG versus non-CpG methylation.

2. Methylation features (e.g. gene bodies) from different contexts (e.g. CpG versus non-CpG) are often provided with the same name. It's easier to deal with these conflicts if we can split the contexts into different modalities, rather than having to figure out some other way to make the features have unique names which may be cumbersome with downstream tools (e.g. running pathway analyses on gene sets).

File: src/scvi/external/methylvi/_module.py (hunk 10 above)
Comment (ethanweinberger) [reply src/scvi/external/methylvi/_module.py:111]: See my reply to your top-level comment for details here. Modalities aren't guaranteed to have the same same shape and we often will have more than one methylation "modality" per cell.

File: src/scvi/external/methylvi/_model.py (hunk 8 above)
Comment (ethanweinberger) [reply src/scvi/external/methylvi/_model.py:547]: I think testing for both hyper/hypo methylation (i.e., larger/smaller) are both interesting.

File: src/scvi/external/methylvi/_model.py (hunk 7 above)
Comment (ethanweinberger) [reply src/scvi/external/methylvi/_model.py:144]: Done!

File: src/scvi/external/methylvi/_model.py (hunk 12 above)
Comment (ethanweinberger) [reply src/scvi/external/methylvi/_model.py:387]: Thanks for the catch- fixed! I hadn't properly accounted for different methylation contexts in this code path previously.

**Comment by ethanweinberger** [comment]
> My suggestion is that we will merge it with the biorxiv version and the tutorials?

Fine with me! It'll probably take a couple of months before we're ready to put the full preprint on biorxiv. That said, if you're planning on releasing scvi-tools 1.2 soon and want to have methylVI ready for that release, I'm happy to merge before the preprint is ready. (Whatever works for you is fine with me :) .)

In the meantime, please let me know what you think of my responses to your latest round of comments.

File: src/scvi/external/methylvi/_model.py (hunk 7 above)
Comment (canergen) [reply src/scvi/external/methylvi/_model.py:143]: I'm missing it?

File: src/scvi/external/methylvi/_module.py (hunk 10 above)
Comment (canergen) [reply src/scvi/external/methylvi/_module.py:115]: Oh, you concatenate along the feature axis (not the cell axis). Makes sense now and thanks for the explaination.

File: docs/user_guide/models/methylvi.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (canergen) [reply docs/user_guide/models/methylvi.md:129]: @ori-kron-wis Can you provide some insight?

File: docs/user_guide/models/methylvi.md (hunk 2 above)
Comment (canergen) [reply docs/user_guide/models/methylvi.md:18]: Missed this one. In that case we can merge as soon as categorical_covariates are added to the codebase. Looks like all else is fixed.

File: src/scvi/external/methylvi/_model.py (hunk 1 above)
Comment (canergen) [reply src/scvi/external/methylvi/_model.py]: I guess the correct way would be to iterate over the layer fields and split the "mod" from above. I don't like treating adata_manager as a bucket. You can have a small helper function that takes care of this before setting up the model and stores it under model.modalities.

File: src/scvi/external/methylvi/_model.py (hunk 8 above)
Comment (canergen) [reply src/scvi/external/methylvi/_model.py:530]: I'll ping you when we have added it there.

File: src/scvi/external/methylvi/_model.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (canergen) [reply src/scvi/external/methylvi/_model.py]: You could do MethylVIModel -> METHYLVI (would be more inline with other models).

**Comment by canergen** [comment]
Can yiu rebase to main? The tests should work now.

File: src/scvi/external/methylvi/_model.py (hunk 7 above)
Comment (ethanweinberger) [reply src/scvi/external/methylvi/_model.py:143]: If you check one line below here I've added the `categorical_covariate_keys` parameter

File: src/scvi/external/methylvi/_model.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ethanweinberger) [reply src/scvi/external/methylvi/_model.py]: In the [current version of the code](https://github.com/scverse/scvi-tools/pull/2834/files#diff-ed595570a5040bd1b9fee7ff3e4ef7902f4ee8d38c2e98b2e72ad32bbc54fad2R38) this has been renamed to METHYLVI as requested

File: src/scvi/external/methylvi/_model.py (hunk 1 above)
Comment (ethanweinberger) [reply src/scvi/external/methylvi/_model.py]: I agree that this initial solution wasn't great. To resolve this, I accessedthe anndata manager within the model's `__init__` function and then retrieved the setup method args to get the modality/context names (see line 84 in the newest version of `_model.py`). Let me know what you think of this solution.

**Comment by ethanweinberger** [comment]
@canergen just responded to your latest round of comments + rebased as requested.

**Comment by ethanweinberger** [comment]
Hey @canergen @ori-kron-wis. Just wanted to check in to see if there are any remaining blockers before this PR can be merged. Thanks!

File: src/scvi/external/methylvi/_model.py [hunk 13]
```diff
@@ -0,0 +1,602 @@
...
+        n_batch = self.summary_stats.n_batch
+        n_cats_per_cov = (
+            self.adata_manager.get_state_registry(REGISTRY_KEYS.CAT_COVS_KEY)[
+                fields.CategoricalJointObsField.N_CATS_PER_KEY
+            ]
+            if REGISTRY_KEYS.CAT_COVS_KEY in self.adata_manager.data_registry
+            else None
+        )
+
+        self.contexts = self.get_anndata_manager(mdata).registry[_SETUP_ARGS_KEY][
```
Comment (canergen) [inline src/scvi/external/methylvi/_model.py:84]: self.get_anndata_manager(mdata, **required=True**)

**Comment by canergen** [comment]
Hi @ethanweinberger thanks for the ping. One tiny comment. What's the status of the tutorial? We would release it with 1.3 but could merge it to main. @ori-kron-wis any blockers from your side?

**Comment by ori-kron-wis** [comment]
We need to fix this pre-commit ruff errors and then we can merge to v1.3

**Comment by ethanweinberger** [comment]
@canergen @ori-kron-wis Thanks for the quick responses :)! I made your suggested change @canergen and fixed the ruff errors @ori-kron-wis.

As for the tutorial: I'm ready to open a PR on the tutorials repository whenever this PR is merged in.

**Comment by ethanweinberger** [comment]
Hey @canergen @ori-kron-wis. Apologies for the spam- just wanted to check if we're good to merge now? I'm ready to open the tutorial PR whenever this gets in.

**Comment by ori-kron-wis** [comment]
I merged. I validated the CUDA test locally (for some reason, the forked branch test did not run in our servers)

**Comment by ethanweinberger** [comment]
Awesome thanks Ori! Will have tutorials ready for you soon.
